algorithms/kernel/dtrees/forest/regression/oneapi/df_regression_train_dense_default_oneapi_impl.i

```diff
@@ -0,0 +1,1245 @@
+
+    const int absentMark    = -1;
+    const int localSize     = _preferableSubGroup;
+    const int nSubgroupSums = _maxLocalSums * localSize < nRows ? _maxLocalSums : (nRows / localSize + !(nRows / localSize));
+
+    auto rowsBuffer        = context.allocate(TypeIds::id<int>(), nRows, &status); // it is filled with marks Present/Absent for each rows
+    auto partialSums       = context.allocate(TypeIds::id<int>(), nSubgroupSums, &status);
+    auto partialPrefixSums = context.allocate(TypeIds::id<int>(), nSubgroupSums, &status);
+    auto totalSum          = context.allocate(TypeIds::id<int>(), 1, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+
+    context.fill(rowsBuffer, absentMark, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+
+#ifdef CHECKTIME
+    timer mark_st = timer::start();
+#endif
+
+    {
+        auto & kernel = kernelMarkPresentRows;
+        KernelArguments args(3);
+        args.set(0, rowsList, AccessModeIds::read);
+        args.set(1, rowsBuffer, AccessModeIds::write);
+        args.set(2, (int)nRows);
+
+        KernelRange local_range(localSize);
+        KernelRange global_range(localSize * nSubgroupSums);
+
+        KernelNDRange range(1);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    {
+        auto & kernel = kernelCountAbsentRowsForBlocks;
+        KernelArguments args(3);
+        args.set(0, rowsBuffer, AccessModeIds::read);
+        args.set(1, partialSums, AccessModeIds::write);
+        args.set(2, (int)nRows);
+
+        KernelRange local_range(localSize);
+        KernelRange global_range(localSize * nSubgroupSums);
+
+        KernelNDRange range(1);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    {
+        auto & kernel = kernelCountAbsentRowsTotal;
+        KernelArguments args(4);
+        args.set(0, partialSums, AccessModeIds::read);
+        args.set(1, partialPrefixSums, AccessModeIds::write);
+        args.set(2, totalSum, AccessModeIds::write);
+        args.set(3, (int)nSubgroupSums);
+
+        KernelRange local_range(localSize);
+        KernelRange global_range(localSize);
+
+        KernelNDRange range(1);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    auto nOOBRowsHost = totalSum.template get<int>().toHost(ReadWriteMode::readOnly);
+
+    nOOBRows = (size_t)nOOBRowsHost.get()[0];
```

Comment:
Check `nOOBRowsHost` on NULL (It critical from scan tool)

Reply:
thanks! Will add.

Reply:
fixed